open_spiel/games/phantom_go.h

```diff
@@ -0,0 +1,179 @@
+// limitations under the License.
+
+#ifndef OPEN_SPIEL_GAMES_PHANTOM_GO_H_
+#define OPEN_SPIEL_GAMES_PHANTOM_GO_H_
+
+#include <array>
+#include <cstring>
+#include <map>
+#include <memory>
+#include <string>
+#include <unordered_set>
+#include <vector>
+
+#include "open_spiel/games/phantom_go/phantom_go_board.h"
+#include "open_spiel/spiel.h"
+#include "open_spiel/spiel_utils.h"
+
+// Game of Phantom Go:
+// https://www.chessprogramming.org/Phantom_Go
+//
+// Parameters:
+//  "komi"              float  compensation for white                  (default = 7.5)
+//  "board_size"        int    rows of the board, usually 9, 13 or 19  (default = 19)
+//  "handicap"          int    number of handicap stones for black     (default = 0)
+//  "max_game_length"   int    maximal lenght of a game                (default = board_size * board_size * 4)
+
+namespace open_spiel {
+namespace phantom_go {
+
+// Constants.
+inline constexpr int NumPlayers() { return 2; }
+inline constexpr double LossUtility() { return -1; }
+inline constexpr double WinUtility() { return 1; }
+inline constexpr int CellStates() { return 3; }  // Black, white, empty.
+
+// Go can only end in a draw when using a round komi.
+// We also treat superko as a draw.
+inline constexpr double DrawUtility() { return 0; }
+
+// All actions must be in [0; NumDistinctActions).
+inline int NumDistinctActions(int board_size) {
+    return board_size * board_size + 1;
+}
+
+// Such high number has been set, mainly because moves on enemy stones are also counted into length
+// And for "clear" resampling, lot of passes and "observation moves" are needed
+inline int DefaultMaxGameLength(int board_size) {
+    return board_size * board_size * 4;
+}
+
+inline int ColorToPlayer(GoColor c) { return static_cast<int>(c); }
+inline GoColor PlayerToColor(Player p) { return static_cast<GoColor>(p); }
+
+// State of an in-play game.
+// Actions are contiguous from 0 to board_size * board_size - 1, row-major, i.e.
+// the (row, col) action is encoded as row * board_size + col.
+// The pass action is board_size * board_size.
+class PhantomGoState : public State {
+ public:
+  // Constructs a Go state for the empty board.
+  PhantomGoState(std::shared_ptr<const Game> game, int board_size, float komi,
+                 int handicap);
+
+  Player CurrentPlayer() const override {
+      return IsTerminal() ? kTerminalPlayerId : ColorToPlayer(to_play_);
+  }
+  std::vector<Action> LegalActions() const override;
+
+  std::array<int, 2> getStoneCount() const;
+
+  std::string ActionToString(Player player, Action action) const override;
+  std::string ToString() const override;
+
+  bool IsTerminal() const override;
+
+  std::unique_ptr<State> ResampleFromMetaposition(
+      int player_id, std::function<double()> rng) const;
+
+  std::unique_ptr<State> ResampleFromInfostate(
+      int player_id, std::function<double()> rng) const;
+
+  std::unique_ptr<State> ResampleFromMetapositionHard(
```

Comment:
Why is there a second method for sampling Metapositions?

Reply:
This method is used, if the ResampleFromMetaposition fails. This can happen if the count of the stones and the situation on the board and in the players observation is in a such coincidence, that randomly picking remaining enemy stones can result in a such situation, that no more enemy stones can be placed. We have agreed, that if the Metaposition resampling is a temporary solution, because Infostate resampling will be implemented, such solution to this problem is applicable.